service/interpreter/timers/greedyTimerProcessor.go

```diff
@@ -0,0 +1,264 @@
+package timers
+
+import (
+	"github.com/indeedeng/iwf/service/interpreter/cont"
+	"github.com/indeedeng/iwf/service/interpreter/interfaces"
+	"time"
+
+	"github.com/indeedeng/iwf/gen/iwfidl"
+	"github.com/indeedeng/iwf/service"
+)
+
+type sortedTimers struct {
+	status service.InternalTimerStatus
+	// Ordered slice of all timers being awaited on
+	timers []*service.TimerInfo
+}
+
+type GreedyTimerProcessor struct {
+	pendingTimers                   sortedTimers
+	stateExecutionCurrentTimerInfos map[string][]*service.TimerInfo
+	staleSkipTimerSignals           []service.StaleSkipTimerSignal
+	provider                        interfaces.WorkflowProvider
+	logger                          interfaces.UnifiedLogger
+}
+
+func NewGreedyTimerProcessor(
+	ctx interfaces.UnifiedContext,
+	provider interfaces.WorkflowProvider,
+	continueAsNewCounter *cont.ContinueAsNewCounter,
+	staleSkipTimerSignals []service.StaleSkipTimerSignal,
+) *GreedyTimerProcessor {
+
+	tp := &GreedyTimerProcessor{
+		provider:                        provider,
+		pendingTimers:                   sortedTimers{status: service.TimerPending},
+		stateExecutionCurrentTimerInfos: map[string][]*service.TimerInfo{},
+		logger:                          provider.GetLogger(ctx),
+		staleSkipTimerSignals:           staleSkipTimerSignals,
+	}
+
+	// start some single thread that manages timers
+	tp.createGreedyTimerScheduler(ctx, continueAsNewCounter)
+
+	err := provider.SetQueryHandler(ctx, service.GetCurrentTimerInfosQueryType, func() (service.GetCurrentTimerInfosQueryResponse, error) {
+		return service.GetCurrentTimerInfosQueryResponse{
+			StateExecutionCurrentTimerInfos: tp.stateExecutionCurrentTimerInfos,
+		}, nil
+	})
+	if err != nil {
+		panic("cannot set query handler")
+	}
+	return tp
+}
+
+func (t *sortedTimers) addTimer(toAdd *service.TimerInfo) {
+
+	if toAdd == nil || toAdd.Status != t.status {
+		panic("invalid timer added")
+	}
+
+	insertIndex := 0
+	for i, timer := range t.timers {
+		if toAdd.FiringUnixTimestampSeconds >= timer.FiringUnixTimestampSeconds {
+			// don't want dupes. Makes remove simpler
+			if toAdd == timer {
+				return
+			}
+			insertIndex = i
+			break
+		}
+		insertIndex = i + 1
+	}
+	t.timers = append(
+		t.timers[:insertIndex],
+		append([]*service.TimerInfo{toAdd}, t.timers[insertIndex:]...)...)
+}
+
+func (t *sortedTimers) removeTimer(toRemove *service.TimerInfo) {
+	for i, timer := range t.timers {
+		if toRemove == timer {
+			t.timers = append(t.timers[:i], t.timers[i+1:]...)
+			return
+		}
+	}
+}
+
+func (t *sortedTimers) pruneToNextTimer(pruneTo int64) *service.TimerInfo {
```

Comment:
looks like it's `prune the pendingSchedules to target time, and also return the earliest time`?
it would be nice add some comment here